Allow suppressing UI blocking dialogs

eugenesvk opened the pull request:
Add 2 user configuration options that would allow to suppress the attention-demanding UI blocking dialogs, instead redirecting warnings/errors to a status message and the full console message 

Simple way to close https://github.com/sublimelsp/LSP/issues/2545 while preserving the current defaults

Though not all modals are affected since some are invoked by the lsp_utils dependency

rchl commented on plugin/tooling.py:
```diff
@@ -130,15 +132,19 @@ def run(self, base_package_name: str) -> None:
         try:
             urllib.parse.urlparse(base_url)
         except Exception:
-            sublime.error_message("The clipboard content must be a URL to a package.json file.")
+            msg = "The clipboard content must be a URL to a package.json file."
+            status = "Clipboard must be a URL to package.json"
+            notify_err(msg, status)
```
I honestly don't think we are gonna accept any of these changes (not in this form). It doesn't make sense to just blindly change all instance of modals. Especially in this file it's a case of explicit user action and it makes sense to report the error in the dialog (the user won't be surprised by it) instead of having to dig for more info the console or potentially miss status field message.

eugenesvk replied:
The users to whom it makes sense to see these kind of dialogs or who worry about missing a status field message can simply continue to use the defaults. Or they can set the regular dialog suppression setting, but leave the error setting on.

So this setting is simply not for them, but for the other group who have never had an issue of missing a status field message (by the way, you could add ❗ symbol if you think this is a very important message, that's harder to miss), don't see the point of a blocking modal for a simple error message. For them these bad interface design patterns don't make sense because they do not reflect the importance of the messages.

> explicit user action and it makes sense to report the error in the dialog (the user won't be surprised by it)

The issue here isn't surprise, but, requiring additional user action for no good reason.

What's your alternative - group/code all the messages and ask the user to provide a list of those that should be suppressed like in lints?

rchl replied:
In this case the error is a direct response to user action so it's fine to present a dialog that can be dismissed easily with a single button press.

If you don't like dialogs so much then maybe just create a plugin that overrides `sublime.error_message` (if that's possible but I think it should)? It's not sensible to expect every package to introduce an option to switch between error dialogs and status field messages.

For some cases I would also prefer to not have dialogs but as discussed in related conversation, there is not necessarily a better way due to ST's limitations.

eugenesvk replied:
> In this case the error is a direct response to user action

Showing a status message is also a direct response

>  so it's fine to present a dialog that can be dismissed easily with a single button press.

It's not fine, you need to provide value to the user to demand action from every single user (this is an **optional** config, disabled by default, so most of the users will continue to suffer through needless actions the way you want them), and in this case you don't - this is simply not an important error message.

Also for some of these cases a blocking message prevents a much better UI: for example, in the next error that appears after the`parse VSCode package JSON` action instead of a blocking `URL must end with 'package.json'"` it'd be better to show the same input field again with the old user value and show a non-blocking error message (e.g., in the status bar or maybe in some popup) so you don't lose valuable user input data and allow for the next user action to be error correction instead of UI unblocking

> If you don't like dialogs so much then maybe just create a plugin that overrides sublime.error_message (if that's possible but I think it should)?

I don't like needless user actionn, not dialogs, so I don't want to override all the error messages, just the ones that aren't important. None in your plugin are, so I'd be fine with overriding all of them.

> It's not sensible to expect every package to introduce an option to switch between error dialogs and status field messages.

I don't expect that, there are plenty of plugins that don't show dialogs, or use less intrusive notifications, and this is a PR for a single package optionally supressing unimportant error messages, not for Sublime core changing everything

jwortmann commented:
Here is my opinion:

- I would probably be okay with such a change to remove (most of) the dialog windows, which don't require user input.
- *if* we want to make this configurable, it should be a single setting and use a better name. Maybe something like `suppress_error_dialogs`. No need to distinguish between `message_dialog` and `error_dialog` imo. But I would probably favor not to add a setting at all for this, unless you want to implement a "don't show this dialog again" checkbox/button.
- the status text should give a reference to the console, like "Failed to start LSP-json - see console for details"
- status message should use the `Window.status_message` method if possible, not `sublime.status_message`. And there are likely a few more things to clean up in the code, e.g. don't use local imports. But I haven't looked too closely yet.
- I don't really like emojis as ⚠️ in the status bar or console
- I found the package for notifications: https://facelessuser.github.io/SubNotify/ but unfortunately it is a separate package and not a dependency. While it could be implemented with a fallback (https://facelessuser.github.io/SubNotify/usage/#tips-and-tricks-for-developers), probably no user has that package installed, so it seems more or less useless to use as it is now. Probably out of scope for this PR, but maybe it could be converted to a regular dependency?

jwortmann commented:
> **Question**: is it possible to make that dependency respect config options from this plugin? Or do dependencies "by design" ignore their clients?

Yes, should be possible. Just load settings via `sublime.load_settings('LSP.sublime-settings')`.

rchl commented on LSP.sublime-settings:
```diff
@@ -49,8 +49,7 @@
 
-  "show_ui_blocking_message": true, // Show UI blocking modals or the full console message and short status notification
-  "show_ui_blocking_err_message": true, // Show UI blocking error modals or the full console message and short status notification
+  "suppress_error_dialogs": false, // Replace UI blocking modals with the full console message and a short status notification
```
- Follow conventions in the file (the way the comment is placed)
- Find a better place for the option - right now it's inserted in the middle of diagnostics-related options.

eugenesvk replied:
moved to others

rchl commented on plugin/code_actions.py:
```diff
@@ -352,7 +353,8 @@ def run_async() -> None:
 
     def _handle_response_async(self, session_name: str, response: Any) -> None:
         if isinstance(response, Error):
-            sublime.error_message(f"{session_name}: {str(response)}")
+            msg = f"{session_name}: {str(response)}"
+            notify_err(msg, msg)
```
Don't use abbreviations like `err`, `msg`.

eugenesvk replied:
renamed

eugenesvk commented:
Updated the PR per the following:

> * I would probably be okay with such a change to remove (most of) the dialog windows, which don't require user input.

switched to suppress by default

> * _if_ we want to make this configurable, it should be a single setting and use a better name. Maybe something like `suppress_error_dialogs`. No need to distinguish between `message_dialog` and `error_dialog` imo. But I would probably favor not to add a setting at all for this, unless you want to implement a "don't show this dialog again" checkbox/button.

Left the setting given that it's suppressed by default, also re the extra button: Sublime still hasn't learned to properly roundtrip user configs, and this button would change files in the User folder, thus breaking subtle formatting, so I'd just leave it in the currently less convenient format

> * the status text should give a reference to the console, like "Failed to start LSP-json - see console for details"

Added a shorter text, otherwise it might not fit

> * status message should use the `Window.status_message` method if possible, not `sublime.status_message`. And there are likely a few more things to clean up in the code, e.g. don't use local imports. But I haven't looked too closely yet.

switched

> * I don't really like emojis as ⚠️ in the status bar or console

They address @rchl concert of users not noticing notifications, ⚠️ make them stand out. I don't mind either way, so removed ⚠️, but left ❗ as I presume those are supposed to still be differentiated somehow

rchl commented:
> > * I don't really like emojis as ⚠️ in the status bar or console
> 
> They address @rchl concert of users not noticing notifications, ⚠️ make them stand out. I don't mind either way, so removed ⚠️, but left ❗ as I presume those are supposed to still be differentiated somehow

I do think that there will be a good portion of users that will not notice something showing up in the status field for a few seconds. Or if they do notice something changing, they might not realize what has changed (if the status field is busy).

So as I said, I don't think that status messages are a good solution in general.

I suppose when user is doing an explicit action then he/she might be more looking for some feedback/result and might pay more attention.

If the user doesn't expect anything then it might be easy to miss status change completely.

eugenesvk commented:
After checking the manually shortened messages, there is actually no ❗ left, only in the short template when the only thing you see is `❗LSP: see console log…` instead of any substatntive message, and since you both don't like them won't adde them to the actual status error messages

> I do think that there will be a good portion of users that will not notice something showing up in the status field for a few seconds

Have you actually tried it? It's harder to miss colored ❗ motion in your vision periphery

> So as I said, I don't think that status messages are a good solution in general.

By the way, for very important messages you could also use these popups, that's in user's focus (at the cursor) (and of course it could fit the whole message)

<img width="245" alt="not" src="https://github.com/user-attachments/assets/096ea33d-3b90-4271-bfaf-41b65b3670e3">

rchl commented on plugin/tooling.py:
```diff
@@ -130,15 +132,19 @@ def run(self, base_package_name: str) -> None:
         except Exception as ex:
-            sublime.error_message(f'Unable to load "{base_url}": {ex}')
+            msg = f'Unable to load "{base_url}": {ex}'
+            notify_error(sublime.active_window(), message, message)
```
Get window from `self.view`

jwortmann commented on plugin/core/logging.py:
```diff
@@ -39,3 +39,33 @@ def exception_log(message: str, ex: Exception) -> None:
+        sublime.message_dialog(message)
+
+
+def notify_error(window: Optional[sublime.Window], message: str, status: str = '❗LSP: see console log…') -> None:
```
I think we don't need 2 separate functions. Current use of `sublime.mesage_dialog` and `sublime.error_message` looks pretty arbitrary. I think all of them are errors, so I'd use only `sublime.error_message`. If the severity is important it could be another parameter `severety`, but I don't think that's necessary. `notify_error` sounds like the appropriate function name, so I would remove the other function.

jwortmann commented on plugin/core/logging.py:
```diff
@@ -39,3 +39,33 @@ def exception_log(message: str, ex: Exception) -> None:
+    if userprefs().suppress_error_dialogs:
+        if window:
+            window.status_message(status)  # print short message to statusbar
+        else:
+            sublime.status_message(status)
```
I assume you used this to satisfy the type checker. I believe window = `None` is only a theoretical case, and if there is no (active) window, I think we don't need to log anyway. So you can just put
```
if not window:
    return
```
at the top of the function.

eugenesvk replied:
done

jwortmann commented on plugin/core/windows.py:
```diff
@@ -295,12 +296,13 @@ def start_async(self, config: ClientConfig, initiating_view: sublime.View) -> No
                 "Re-enable by running \"LSP: Enable Language Server In Project\" from the Command Palette.",
                 "\n\n--- Error: ---\n{1}"
             )).format(config.name, str(e))
+            status = f"LSP: Failed to start {config.name}… See console"
```
Do we need the `LSP:` prefix in the status messages? It should be obvious if there is either the `config.name` or when it was something directly triggered by the user.

eugenesvk replied:
ok, there is also the bigger console message with more detailed errors to fall back to

eugenesvk commented on plugin/core/logging.py:
```diff
@@ -39,3 +39,33 @@ def exception_log(message: str, ex: Exception) -> None:
+    from .settings import userprefs
+    if userprefs().suppress_error_dialogs:
+        if window:
+            window.status_message(status)  # print short message to statusbar
```
removed, but the non-obvious part was to differentiate status_message vs print (short vs full)

eugenesvk commented on plugin/core/logging.py:
```diff
@@ -39,3 +39,33 @@ def exception_log(message: str, ex: Exception) -> None:
     print(prefix + ":", *args)
+
+
+def notify(window: Optional[sublime.Window], message: str, status: str = 'LSP: see console log…') -> None:
```
renamed, but the defaults are for those future cases

jwortmann replied:
When it's used in the future, I'd expect that you still pass a message to this function and don't use that default text.

How about `status_message: str | None = None` and then
```python
if status_message is None:
    status_message = message
```
because a few times in the code you used the same console and status message.

Also I think now that for the console message we should add the `LSP:` prefix, to make it easy to see which lines come from LSP and because the space is not so limited in the console.

eugenesvk replied:
ok, added print prefix, deduped and made it optional

eugenesvk commented on plugin/core/logging.py:
```diff
@@ -39,3 +39,33 @@ def exception_log(message: str, ex: Exception) -> None:
+    """Pick either of the 2 ways to show a user notification message:
+      - via a detailed console message and a short status message
+      - via a blocking modal dialog"""
+    from .settings import userprefs
```
yes, circular import was the reason, moved these to a separate file

jwortmann commented on plugin/core/logging_notify.py:
```diff
@@ -0,0 +1,20 @@
+
+
+def notify_error(window: sublime.Window | None, message: str, status_message: str = 'LSP: see console log…',
+                 severe: bool = True) -> None:
```
You never used that argument in the function calls, so I think we don't need that and it can be removed. The difference between `sublime.error_message` and `sublime.message_dialog` is just an additional icon anyways (at least on Windows), so it's probably not very important.

eugenesvk replied:
k, removed

I've never used it because you said that all current messages are the same, so they all got bumped to "severe" in the process of converting to a single function.

I haven't reviewed all the messages to see whether a different type is justified. This again was more future-proof in case some messages will be deemed less severe

jwortmann replied:
Yeah but if we need it in the future, we can simply add it then. I'm not gonna say that I have never done this myself, but in general it's probably not very useful to add unused code with the reasoning that it could potentionally be used in the future.

eugenesvk commented on sublime-package.json:
```diff
@@ -449,6 +449,11 @@
             },
+            "suppress_error_dialogs": {
+              "type": "boolean",
+              "default": false,
```
fixed

eugenesvk replied:
>  I would just print the messages to the status bar and console.

So let's revert it back to true then here and in the settings

eugenesvk commented:
> Perhaps as a compromise the error messages from the tooling / troubleshoot command

But user configuration is that compromise. I, for one, would never want to see these as blocking either, showing informational error messages doesn't warrant this. And this package then can define whatever compromising defaults you want, it won't affect others permanently

Besides this goes back to the 2-tier error message system that was just removed, where you had the flexibility of suppressing dialogs of only 1 type, both at the level of the plugin, and individual user
